leverage LRU caching for scheduled checks and expr/rule tests

Dieterbe opened the pull request:
* for executing expressions/rules via the web UI, we use
a cache that we retain during the lifetime of bosun
Matt and I decided not to expire the cache at given points (such as
reloading rule page), but I forgot why. ?
the only risk is that if you query your store for data -5m to now
and your store doesn't have the latest points up to date,
and then 5m from now you query -10min to -5m you'll get the same
cached data, including the incomplete last points

* for scheduled alert checking, the cache lifetime is only each
specific run so that different queries of the same
data for different templates are optimized and values reused in
templates are guaranteed to be in sync.
we don't make the lifetime longer to prevent affecting
future queries (perhaps a band() a week
from now which queries now-current data we just queried and might be
incomplete)
also there's likely less benefit (low hit rate) in keeping the cache
across alerting runs

note that Schedule.Check(), Schedule.Poll() and
Schedule.CheckUnknown() currently use their own caches for each run.
Schedule.Poll() runs on its own schedule and Schedule.CheckUnknown()
runs every checkinterval/4..
We should probably reuse the caching more across these checking
functions.  But I leave this up to Matt who knows more about this.

Note:
* graphite is tested, opentsdb is not.
* i did not work on logstash

Dieterbe commented:
i realized i forgot to turn the lru/singleflight deps into vendored deps. will do that

Dieterbe commented:
because some changes had been applied to tsdb.go in master, I rebased it to incorporate those changes.

a notable omission in this PR is knobs exposed to the admin to tune the caching. because i wasn't quite sure how to go about this.
right now LRU is hardcoded to 100 items.  what i suggest is 2 config options (i couldn't come up with better names, help me out!):
```
lruSizeRun # to limit lru size of the cache created for every scheduled alerting run
lruSizeInterface # to limit lru size of the long running cache that serves the web interface
```
thoughts? @kylebrandt @mjibson

kylebrandt commented:
I'm guessing less than 10% of users will know what LRU means - I would call
alertCache and interfaceCache - or something that makes more sense from a
user perspective.

Excited about this PR!

On Fri, Feb 13, 2015 at 11:21 AM, Dieter Plaetinck <notifications@github.com
> wrote:

> because some changes had been applied to tsdb.go in master, I rebased it
> to incorporate those changes.
>
> a notable omission in this PR is knobs exposed to the admin to tune the
> caching. because i wasn't quite sure how to go about this.
> right now LRU is hardcoded to 100 items. what i suggest is 2 config
> options (i couldn't come up with better names, help me out!):
>
> lruSizeRun # to limit lru size of the cache created for every scheduled alerting run
> lruSizeInterface # to limit lru size of the long running cache that serves the web interface
>
> thoughts? @kylebrandt <https://github.com/kylebrandt> @mjibson
> <https://github.com/mjibson>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bosun-monitor/bosun/pull/715#issuecomment-74280095>.
>

mjibson commented on cache/cache.go:
```diff
@@ -0,0 +1,43 @@
+
+func New() *Cache {
+	return &Cache{
+		lru: lru.New(100),
```
The size should be specifiable in New.

Dieterbe replied:
agreed.  i didn't do this yet because how to specify the config for this is still an open question.

mjibson replied:
Ok. This can wait until a later PR then. That's fine.

mjibson commented on cache/cache.go:
```diff
@@ -0,0 +1,43 @@
+	lru *lru.Cache
+}
+
+type Result struct {
```
No need to export this.

mjibson commented on cache/cache.go:
```diff
@@ -0,0 +1,43 @@
+	// our lock only serves to protect the lru.
+	// we can (and should!) do singleflight requests concurently
+	val, err := c.g.Do(key, getFn)
+	r := Result{val, err}
```
&Result, I think

mjibson commented on cache/cache.go:
```diff
@@ -0,0 +1,43 @@
+	result, ok := c.lru.Get(key)
+	c.Unlock()
+	if ok {
+		res := result.(Result)
```
(*Result)

mjibson commented on cache/cache.go:
```diff
@@ -0,0 +1,43 @@
+	val, err := c.g.Do(key, getFn)
+	r := Result{val, err}
+	c.Lock()
+	c.lru.Add(key, r)
```
This has the potential to dogpile adding entries to the cache, including locking c.lru. What about moving c.lru.Add into getFn? For example:

```
val, err := c.g.Do(key, func() (interface{}, error) {
  v, err := getFn()
  c.lru.Add(key, &Result{v, err})
  return v, err
})
```

And remove the c.Lock/Unlock calls.

Dieterbe replied:
in practice i don't think this would be an issue, and in current context the lock contention will be minimal.
however your approach is more correct and allows the cache to be used in high-request scenarios, so i'll adjust.

frankly it's a little messy that the lock is on global level, so we release lock after checking (we don't want to hold a global lock during the getFn() call) but this also allows for race conditions. (object appearing in cache right after we checked, and hence we still call getFn() needlessly).  not an issue in the current context, but could be done more correctly

mjibson commented on cmd/bosun/conf/conf.go:
```diff
@@ -67,20 +68,20 @@ type Conf struct {
 // TSDBCacheContext returns an OpenTSDB context with caching and limited to
 // c.ResponseLimit. A nil context is returned if TSDBHost is not set.
-func (c *Conf) TSDBCacheContext() opentsdb.Context {
+func (c *Conf) TSDBCacheContext(cache cache.Cache) opentsdb.Context {
```
These should all be *cache.Cache. Same in other files.

mjibson commented on cmd/bosun/sched/check.go:
```diff
@@ -191,7 +193,8 @@ func (s *Schedule) RunHistory(r *RunHistory) {
 		log.Println("checkUnknown")
-		r := s.NewRunHistory(time.Now())
+		cache := cache.New()
+		r := s.NewRunHistory(time.Now(), *cache)
```
Just merge these two lines

mjibson commented on cmd/bosun/sched/check.go:
```diff
@@ -67,7 +68,8 @@ func (s *Schedule) Check(T miniprofiler.Timer, now time.Time) (time.Duration, er
 	}
-	r := s.NewRunHistory(now)
+	cache := cache.New()
+	r := s.NewRunHistory(now, *cache)
```
merge these lines

mjibson commented on cmd/bosun/sched/notify.go:
```diff
@@ -14,7 +15,8 @@ import (
 	for {
-		rh := s.NewRunHistory(time.Now())
+		cache := cache.New()
+		rh := s.NewRunHistory(time.Now(), *cache)
```
merge

mjibson commented on cmd/bosun/web/expr.go:
```diff
@@ -14,12 +14,23 @@ import (
 	"bosun.org/opentsdb"
 )
 
+var cacheObj cache.Cache
```
var cacheObj = cache.New()

mjibson commented on graphite/graphite.go:
```diff
@@ -2,6 +2,7 @@
 package graphite // import "bosun.org/graphite"
 
 import (
+	"bosun.org/cache"
```
This should not be grouped with the stdlib imports. Move it down with a newline separating

mjibson replied:
Actually, this package should not depend on cache at all. Caching should be at a higher level.

mjibson commented on opentsdb/tsdb.go:
```diff
@@ -2,6 +2,7 @@
 package opentsdb // import "bosun.org/opentsdb"
 
 import (
+	"bosun.org/cache"
```
Both this and the graphite packages should instead have their cache managed in their query functions in funcs.go (timeTSDBRequest, timeGraphiteRequest). That way the low levels packages don't have to know about the cache at all.

Dieterbe replied:
does that mean throwing out opentsdb's CacheContext ? graphite and opentsdb would each just have a simple context that does querying (and perhaps filtering and limiting) ?

mjibson replied:
Correct.

Dieterbe replied:
ok that's an interesting design. so basically it's up to the thing that calls opentsdb/graphite queries to have their own cache before invoking queries.  But there can be multiple things invoking tsdb/graphite.  I'll give this a shot and see how I can make this work.

Dieterbe commented:
i addressed all comments. some thoughts:

1. timeGraphiteRequest and timeTSDBRequest now do their own cache handling, as requested. the change saves about 16 lines but for a cache object to reach that point, it still has to travel through a bunch of a function calls, which i think is unavoidable
2. the "dumb" opentsdb.Host Context might be unneeded. maybe we should get rid of it, and of the Context interface, and rename LimitContext to Context.
3. we should document (and perhaps revisit) what RunHistory and expr.State are supposed to be.  we have some calls with a lot of args (like `func (e *Expr) Execute`), i'ld like to address this and clean this up at some point (i.e. pass around all contexts and the cache obj together in 1 struct)
i currently use runhistory and state to contain a reference to the cache,
4. Schedule.Poll() and Schedule.CheckUnknown() both create runhistories to do their thing (and instantiate a new cache), although they don't seem to execute any queries to the backend.  if this is the idea they probably should not create a cache.  if not, should the cache be shared with the Schedule.Check() runs? 
5. /api/egraph/{bs}.svg (ExprGraph), /api/expr (Expr) and /api/rule (Rule, actually procRule) all share a cache throughout the lifetime of bosun.i'm not sure this is a good idea and i forgot why we decided to never expire this cache.
6. notably "/api/graph" (Graph() in web package) queries opentsdb directly. would it make sense to use the cache here? perhaps control this via url options? are the opentsdb it takes in the same format as elsewhere (for cache keys to work properly)
7. sched/template/Context calls eval which uses a runhistory to execute expr.Execute.  i haven't figured out yet how this context works, does it have the same RunHistory (and hence cache) as the alerting run we're in?
8. in conf.go we set up Funcs["alert"] which will call expr.ExecuteState when called. not sure when/how this gets called and if the appropriate cache will be used.

i've sketched a diagram of the call graph and where caches are created and shared. if interested i can take a pic and share it.

mjibson commented on graphite/graphite.go:
```diff
@@ -26,6 +26,10 @@ type Series struct {
 type DataPoint []json.Number
 
+func (r *Request) CacheKey() string {
+	return fmt.Sprintf("graphite-%d-%d-%s", r.Start.Unix(), r.End.Unix(), r.Targets)
```
How about just convert r to JSON, or r.Targets to JSON to avoid potential encoding collisions.

mjibson commented on cmd/bosun/web/chart.go:
```diff
@@ -198,7 +198,11 @@ func ExprGraph(t miniprofiler.Timer, w http.ResponseWriter, r *http.Request) (in
 		return nil, fmt.Errorf("egraph: requires an expression that returns a series")
 	}
-	res, _, err := e.Execute(schedule.Conf.TSDBCacheContext(), schedule.Conf.GraphiteContext(), schedule.Conf.LogstashElasticHost, t, now, autods, false, schedule.Search, nil)
+	// TODO does schedule.Conf ever change? maybe we can tie these contexts to this package and reuse them
```
Yes, it could. So these should be generated each time as they are here.

mjibson commented on cmd/bosun/web/expr.go:
```diff
@@ -29,7 +36,11 @@ func Expr(t miniprofiler.Timer, w http.ResponseWriter, r *http.Request) (interfa
 		return nil, err
 	}
-	res, queries, err := e.Execute(schedule.Conf.TSDBCacheContext(), schedule.Conf.GraphiteContext(), schedule.Conf.LogstashElasticHost, t, now, 0, false, schedule.Search, nil)
+	// TODO does schedule.Conf ever change? maybe we can tie these contexts to this package and reuse them
```
Same as above.

Dieterbe commented:
addressed all remaining issues, amended everything into 1 commit.
With Matt I went over all those points above, and basically we agree that some cleanup/refactoring should happen, but that can wait.
Schedule.Poll() and Schedule.CheckUnknown() don't need a cache indeed, Poll doesn't even need a RH anymore (to be cleaned up in #756 ). the cases where i wasn't sure the logic is correct, well, we agreed that all the logic should be correct :)

to keep this moving, I propose to keep the following items for a followup PR:
* tunable config knows for the LRU cache
* "/api/graph" (Graph() in web package) leverage caching
* caching for logstash/ES

I can put the first on my plate. not sure about 2/3.

Dieterbe commented:
just pushed a 2nd commit to fix 756 as well.

mjibson commented on cmd/bosun/sched/check.go:
```diff
@@ -67,7 +70,7 @@ func (s *Schedule) Check(T miniprofiler.Timer, now time.Time) (time.Duration, er
 		return 0, fmt.Errorf("check already running")
 	}
-	r := s.NewRunHistory(now)
+	r := s.NewRunHistory(now, cache.New(100))
```
This should be much higher. Maybe even math.MaxInt.

Dieterbe replied:
fixed

mjibson commented on cmd/bosun/cache/cache.go:
```diff
@@ -0,0 +1,48 @@
+
+func (c *Cache) Get(key string, getFn func() (interface{}, error)) (interface{}, error) {
+	// for simplicity we use a global lock on the LRU
+	// this might cause races where an object appears right after finding it doesn't exist
```
This line and the one below are no longer true since we changed the locking in the get fn.

mjibson commented on cmd/bosun/cache/cache.go:
```diff
@@ -0,0 +1,48 @@
+import (
+	"bosun.org/_third_party/github.com/golang/groupcache/lru"
+	"bosun.org/_third_party/github.com/golang/groupcache/singleflight"
+	"sync"
```
These are not sorted correctly. Use goimports to fix.

Dieterbe replied:
cool tip. fixed
